lib/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+use std::{
+    ffi::{CStr, CString},
+    path::Path,
+    str::FromStr,
+    sync::Once,
+};
+
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("error when calling quilc_compile_quil: {0}")]
+    CompileQuil(String),
+    #[error("error when calling quilc_compile_protoquil: {0}")]
+    CompileProtoquil(String),
+    #[error("error when calling quilc_parse_quil: {0}")]
+    ParseQuil(String),
+    #[error("program string contained unexpected NUL character: {0}")]
+    UnexpectedNul(#[from] std::ffi::NulError),
+    #[error("error when calling quilc_build_nq_linear_chip: {0}")]
+    BuildNqLinearChip(String),
+    #[error("error when calling quilc_parse_chip_spec_isa_json: {0}")]
+    ParseChip(String),
+    #[error("error when calling quilc_print_program: {0}")]
+    PrintProgram(String),
+    #[error("error when calling quilc_program_string: {0}")]
+    ProgramString(String),
+    #[error("invalid UTF-8 program: {0}")]
+    ProgramUtf8(#[from] std::str::Utf8Error),
+}
+
+static START: Once = Once::new();
+
+/// Initializes libquilc using it's core image. No-op after the first call.
+fn init_libquilc() {
+    START.call_once(|| {
+        let path = match std::env::var("LIBQUILC_CORE_PATH") {
+            Ok(path) => path,
+            Err(_) => "libquilc.core".to_string(),
+        };
+        if !Path::new(&path).exists() {
+            // TODO Make this an error rather than a panic
+            panic!("Could not find libquilc core file. Do you need to set LIBQUILC_CORE_PATH environment variable?");
+        }
+        let bytes = match std::env::var("LIBQUILC_CORE_PATH") {
+            Ok(mut path) => {
+                path.push('\0');
+                path.as_bytes().to_vec()
+            }
+            Err(_) => b"libquilc.core\0".to_vec(),
+        };
+        let mut c_chars: Vec<i8> = bytes.iter().map(|c| *c as i8).collect();
+        let ptr = c_chars.as_mut_ptr();
+        unsafe {
+            init(ptr);
+        }
+    })
+}
+
+/// A quilc chip specification
+#[derive(Clone, Debug)]
+pub struct Chip(chip_specification);
+
+unsafe impl Send for Chip {}
```

Comment:
These are required to play nicely with pyo3.

Reply:
What is `chip_specification, and how do we know it's safe to `Send`?

Reply:
A pointer to memory (which is not mutated) used by quilc.

Reply:
`unsafe` code should always have a comment explaining why it's valid, I think; could you please add a comment (here and on the other `unsafe impl`) explaining that the underlying memory is not expected to ever be mutated (at least while Python can hold a pointer to it)?